onnxruntime/core/providers/cuda/nn/rms_norm.cc

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/cuda/nn/rms_norm.h"
+#include "core/providers/cuda/nn/layer_norm.h"
+#include "core/providers/cuda/nn/layer_norm_impl.h"
+#include "core/providers/cuda/cuda_common.h"
+#include "core/providers/cpu/nn/layer_norm_helper.h"
+
+namespace onnxruntime {
+namespace cuda {
+
+// RMSNorm uses LayerNorm kernel, which only supports X and scale both
+// being the same data type.
+#define REGISTER_KERNEL_TYPED(T)                                                              \
+  ONNX_OPERATOR_TYPED_KERNEL_EX(RMSNormalization, kOnnxDomain, 23, T, kCudaExecutionProvider, \
+                                (*KernelDefBuilder::Create())                                 \
+                                    .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())    \
+                                    .TypeConstraint("V", DataTypeImpl::GetTensorType<T>()),   \
+                                RMSNorm<T, float, T, true>);
+
+REGISTER_KERNEL_TYPED(float)
+REGISTER_KERNEL_TYPED(double)
+REGISTER_KERNEL_TYPED(MLFloat16)
+REGISTER_KERNEL_TYPED(BFloat16)
+
+// The following code is shared from "core/providers/cuda/nn/layer_norm.cc".
+// It is used to implement the RMSNorm kernel, which is a simplified version of LayerNorm.
+template <typename T, typename U, typename V, bool simplified>
+RMSNorm<T, U, V, simplified>::RMSNorm(const OpKernelInfo& op_kernel_info)
+    : CudaKernel(op_kernel_info) {
+  ORT_ENFORCE(op_kernel_info.GetAttr("axis", &axis_).IsOK());
+  float tmp_epsilon;
+  ORT_ENFORCE(op_kernel_info.GetAttr<float>("epsilon", &tmp_epsilon).IsOK());
+  epsilon_ = tmp_epsilon;
+}
+
+template <typename T, typename U, typename V, bool simplified>
+Status RMSNorm<T, U, V, simplified>::ComputeInternal(OpKernelContext* ctx) const {
+  typedef typename ToCudaType<T>::MappedType CudaT;
+  typedef typename ToCudaType<U>::MappedType CudaU;
+  typedef typename ToCudaType<V>::MappedType CudaV;
+  // Inputs
+  const Tensor* X = ctx->Input<Tensor>(0);
+  const Tensor* scale = ctx->Input<Tensor>(1);
+  const Tensor* bias = ctx->Input<Tensor>(2);
```

Comment:
I do not see bias input in op definition:
https://onnx.ai/onnx/operators/onnx__RMSNormalization.html